### DISCUSSION TITLE
Configuring cabal to tell HPC to ignore certain functions

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [Copilot-Language/copilot#554] [Q&A]
I want to configure a cabal package (specifically, `copilot-core`) so that, when HPC is executed during testing with coverage enabled, it ignores certain definitions, for example:

- Anything that is auto-generated (e.g., `Show`).
- Data constructors and field accessors, for which GHC also defines functions automatically, are all assumed to work.
- `Proxy` doesn't need to be WHNFed.

I think there's a way to extract the tix file and modify it, but I'm trying to make this automated, and configure the cabal package so that the right options are passed by cabal to HPC, and that way even hackage will pick it up when it reports the coverage of our library.

Any pointers?

Also posted at: https://www.reddit.com/r/haskell/comments/1g8mm82/configuring_cabal_to_tell_hpc_to_ignore_certain/

**Comment by ivanperez-keera** [comment]
Follow-up: I managed to get over 90%, but I'm still finding it hard to hit some expressions. My current branch is: https://github.com/ivanperez-keera/copilot/tree/develop-copilot-core-coverage.

This is what I'm seeing:

![image](https://github.com/user-attachments/assets/e97e19e9-c8dd-4544-a2c9-2847854efda0)

Some cases are due to laziness:

![image](https://github.com/user-attachments/assets/08f32b77-0db0-4f55-ab31-55f58fd4dc95)

In other cases, I don't understand why I can't hit the expression. For example:

![image](https://github.com/user-attachments/assets/c6fb3183-1b91-4765-8db5-4ff1d7e1b639)

I thought this would be explored fully during comparison of values generated using `typeOf` using equality, but it suggests that it never looks at the `typeOf` inside an array.

Help to improve the tests to increase coverage would be appreciated.

**Reply by RyanGlScott** [reply]
I think the reason that the `typeOf` argument to `Array` is never evaluated is because `Array` stores types in two ways:

```hs
data Type :: * -> * where
  Array :: forall n t. (KnownNat n, Typed t) => Type t -> Type (Array n t)
```

Namely, it stores both an implicit `Typed t` constraint as well as an explicit `Type t` field. It appears that the surrounding code is only making use of the `Typed t` constraint without making use of the `Type t` field.

One proposal: what if we got rid of either the `Typed t` constraint or the `Type t` field? I claim that having both is redundant, as you can always convert from a `Typed t` constraint to a `Type t` value (and vice versa). My belief is that the coverage issue related to the `Typed (Array n t)` instance would go away if we did this.

**Reply by RyanGlScott** [reply]
Actually, there might be a simpler solution. Are you certain that the `copilot-core` test suite actually makes meaningful use of the `Typed (Array n t)` instance? I suspect that the call sites to `typeOf` are only evaluating the `Array` data constructor, but not the underlying field. For instance, if I make the following quick-and-dirty change to the `copilot-core` test suite:

```diff
diff --git a/copilot-core/tests/Test/Copilot/Core/Type.hs b/copilot-core/tests/Test/Copilot/Core/Type.hs
index 9f9cb6d5..ccce150d 100644
--- a/copilot-core/tests/Test/Copilot/Core/Type.hs
+++ b/copilot-core/tests/Test/Copilot/Core/Type.hs
@@ -253,7 +253,7 @@ utypes =
     ]
   where
     a1 :: Type (Array 3 Int8)
-    a1 = Array Int8
+    a1 = typeOf
 
     a2 :: Type (Array 4 Int8)
     a2 = Array Int8
```

Then I do see the `typeOf` in `typeOf = Array typeOf` become evaluated in the coverage report.

**Comment by ivanperez-keera** [comment]
There is now an issue in HPC to track this: https://gitlab.haskell.org/hpc/hpc-bin/-/issues/21
